pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/tenant"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+// LogResultCacheMetrics is the metrics wrapper used in log result cache.
+type LogResultCacheMetrics struct {
+	CacheHit  prometheus.Counter
+	CacheMiss prometheus.Counter
+}
+
+// NewLogResultCacheMetrics creates metrics to be used in log result cache.
+func NewLogResultCacheMetrics(registerer prometheus.Registerer) *LogResultCacheMetrics {
+	return &LogResultCacheMetrics{
+		CacheHit: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_hit_total",
+		}),
+		CacheMiss: promauto.With(registerer).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "query_frontend_log_result_cache_miss_total",
+		}),
+	}
+}
+
+// NewLogResultCache creates a new log result cache middleware.
+// Currently it only caches empty filter queries, this is because those are usually easily and freely cacheable.
+// Log hits are difficult to handle because of the limit query parameter and the size of the response.
+// In the future it could be extended to cache log hits.
```

Comment:
Maybe that is clearer so a "log hit" is not confused with a "log result cache hit".

```suggestion
// In the future it could be extended to cache non-empty query results.
```